plugins/inputs/ping_native/ping.go

```diff
@@ -0,0 +1,260 @@
+package ping
+
+import (
+	"errors"
+	"net"
+	"runtime"
+	"time"
+
+	ping "github.com/glinton/ping"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+type Ping struct {
+	PingInterval float64  `toml:"ping_interval"` // Interval at which to ping (ping -i <INTERVAL>)
+	Count        int      `toml:"count"`         // Number of pings to send (ping -c <COUNT>)
+	Timeout      float64  `toml:"timeout"`       // Per-ping timeout, in seconds. 0 means no timeout (ping -W <TIMEOUT>)
+	Deadline     int      `toml:"deadline"`      // Ping deadline, in seconds. 0 means no deadline. (ping -w <DEADLINE>)
+	Interface    string   `toml:"interface"`     // Interface or source address to send ping from (ping -I/-S <INTERFACE/SRC_ADDR>)
+	Hosts        []string `toml:"hosts"`         // Hosts to ping
+	IPV6         bool     `toml:"ipv6"`          // Whether to ping ipv6 addresses
+
+	network    string                 // network is the network to listen on ("udp4", "udp6", "ip4:icmp", "ip6:ip6-icmp")
+	listenAddr string                 // listenAddr is the address associated with the interface defined.
+	hostCache  []net.Addr             // hosts to ping
+	hosts      map[string]string      // ip to host map
+	rcvdCache  map[string]pingResults // cache of echo responses received
+}
+
+func (*Ping) Description() string {
+	return "Ping given host(s) and return statistics"
+}
+
+const sampleConfig = `
+  ## List of hosts to ping.
+  hosts = ["8.8.8.8"]
+
+  ## Number of pings to send per collection.
+  # count = 1
+
+  ## Interval, in s, at which to ping.
+  # ping_interval = 1.0
+
+  ## Per-ping timeout, in s (0 == no timeout).
+  # timeout = 1.0
+
+  ## Total-ping deadline, in s. Set to value equal to or lower than agent interval.
+  # deadline = 10
+
+  ## Interface or source address to send ping from.
+  # interface = ""
+
+	## Whether to ping ipv6 addresses.
+  # ipv6 = false
+`
+
+func (*Ping) SampleConfig() string {
+	return sampleConfig
+}
+
+func getAddr(iface string) string {
+	if addr := net.ParseIP(iface); addr != nil {
+		return addr.String()
+	}
+
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return ""
+	}
+
+	var ip net.IP
+	for i := range ifaces {
+		if ifaces[i].Name == iface {
+			addrs, err := ifaces[i].Addrs()
+			if err != nil {
+				return ""
+			}
+			if len(addrs) > 0 {
+				switch v := addrs[0].(type) {
+				case *net.IPNet:
+					ip = v.IP
+				case *net.IPAddr:
+					ip = v.IP
+				}
+				if len(ip) == 0 {
+					return ""
+				}
+				return ip.String()
+			}
+		}
+	}
+
+	return ""
+}
+
+func (p *Ping) buildHostCache(acc telegraf.Accumulator) {
+	if p.hostCache != nil {
+		return
+	}
+
+	p.network = "ip4:icmp"
+	if runtime.GOOS == "darwin" || runtime.GOOS == "linux" {
```

Comment:
I think it could be nice to allow the selection of the type of socket you want to create or to implement an algorithm like iputils does with fallback.  I'm also 90% sure modern BSD will support non-privileged ping sockets, though not sure if this is a limitation of icmp or maybe improperly documented.

What if on non-Windows platform we try ping sockets, and then try raw sockets, and then error?

https://github.com/iputils/iputils/blob/master/ping.c#L109